### PR TITLE
feat: ASGI middleware

### PR DESCRIPTION
**Pull request by untitaker**
Diff to sentry-asgi:

<details><summary>click me</summary>

```diff
diff --git a/../sentry-asgi/sentry_asgi/middleware.py b/sentry_sdk/integrations/asgi.py
index 37d1117..4c40750 100644
--- a/../sentry-asgi/sentry_asgi/middleware.py
+++ b/sentry_sdk/integrations/asgi.py
@@ -1,35 +1,44 @@
 import functools
 import urllib

-import sentry_sdk
-from sentry_sdk.utils import event_from_exception, exc_info_from_error
+from sentry_sdk._types import MYPY
+from sentry_sdk.hub import Hub, _should_send_default_pii
+from sentry_sdk.integrations._wsgi_common import _filter_headers
+from sentry_sdk.utils import transaction_from_function

+if MYPY:
+    from typing import Dict
+
+
+class SentryAsgiMiddleware:
+    __slots__ = ("app",)

-class SentryMiddleware:
     def __init__(self, app):
         self.app = app

     async def __call__(self, scope, receive, send):
-        hub = sentry_sdk.Hub.current
-        with sentry_sdk.Hub(hub) as hub:
+        hub = Hub.current
+        with Hub(hub) as hub:
             with hub.configure_scope() as sentry_scope:
+                sentry_scope._name = "asgi"
                 processor = functools.partial(self.event_processor, asgi_scope=scope)
                 sentry_scope.add_event_processor(processor)
-                try:
-                    await self.app(scope, receive, send)
-                except Exception as exc:
-                    hub.capture_exception(exc)
-                    raise exc from None
+
+            try:
+                await self.app(scope, receive, send)
+            except Exception as exc:
+                hub.capture_exception(exc)
+                raise exc from None

     def event_processor(self, event, hint, asgi_scope):
         if asgi_scope["type"] in ("http", "websocket"):
             event["request"] = {
                 "url": self.get_url(asgi_scope),
                 "method": asgi_scope["method"],
-                "headers": self.get_headers(asgi_scope),
+                "headers": _filter_headers(self.get_headers(asgi_scope)),
                 "query_string": self.get_query(asgi_scope),
             }
-        if asgi_scope.get("client"):
+        if asgi_scope.get("client") and _should_send_default_pii():
             event["request"]["env"] = {"REMOTE_ADDR": asgi_scope["client"][0]}
         if asgi_scope.get("endpoint"):
             event["transaction"] = self.get_transaction(asgi_scope)
@@ -66,7 +75,7 @@ class SentryMiddleware:
         """
         Extract headers from the ASGI scope, in the format that the Sentry protocol expects.
         """
-        headers = {}
+        headers = {}  # type: Dict[str, str]
         for raw_key, raw_value in scope["headers"]:
             key = raw_key.decode("latin-1")
             value = raw_value.decode("latin-1")
@@ -80,12 +89,4 @@ class SentryMiddleware:
         """
         Return a transaction string to identify the routed endpoint.
         """
-        endpoint = scope["endpoint"]
-        qualname = (
-            getattr(endpoint, "__qualname__", None)
-            or getattr(endpoint, "__name__", None)
-            or None
-        )
-        if not qualname:
-            return None
-        return "%s.%s" % (endpoint.__module__, qualname)
+        return transaction_from_function(scope["endpoint"])
```

</details>

### CODE REVIEW

**Comment by untitaker** [comment]
@tomchristie how would you like to be credited?

* new line in LICENSE
* note in ASGI docs (docs.sentry.io)
* code comment

choose `n`

**Comment by tomchristie** [comment]
Not fussed - whichever’s easiest.

**Comment by untitaker** [comment]
I will go for module docstring + docs.sentry.io

**Comment by tomchristie** [comment]
If you ping me when there's a release with this, then I'll update any relevant places to point towards the `sentry-python` implementation, and punt the `encode` one into deprecated mode.

File: sentry_sdk/integrations/asgi.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, scope, receive=None, send=None):
+        if receive is None or send is None:
+
+            async def run_asgi2(receive, send):
+                return await self._run_app(
+                    scope, lambda: self.app(scope)(receive, send)
```
Comment (untitaker) [inline sentry_sdk/integrations/asgi.py]: @tomchristie does this look legit to you? Am I supposed to call `app(scope)` eagerly?

**Pull request merged**